### PR TITLE
BUG: Fix problem with Scipy.integrate.solve_bvp for big problems

### PR DESCRIPTION
**Pull request by fetorres**
This commit solves fatal errors when solve_bvp is used for a big problem, i.e. size > 50,
and the unknowns include parameters that need solving.
This commit includes tests, with and without Jacobians for both the function and the boundary conditions.
 The new tests adapt already existing tests to make them easier to compare and understand.

### CODE REVIEW

File: scipy/integrate/tests/test_bvp.py [hunk 1]
```diff
@@ -532,6 +603,41 @@ def test_big_problem():
...
+                            fun_jac=fun_jac, bc_jac=bc_jac)
+
+            assert_equal(sol.status, 0)
+            assert_(sol.success)
+
+            assert_allclose(sol.p, [1, 1], rtol=1e-4)
+
+            sol_test = sol.sol(x_test)
+
+            for isol in range(0, n, 4):
```
Comment (nmayorov) [inline scipy/integrate/tests/test_bvp.py:624]: Maybe this check can be implemented more straightforward. But it's fine as it is.

File: scipy/integrate/tests/test_bvp.py [hunk 2]
```diff
@@ -532,6 +603,41 @@ def test_big_problem():
...
+                assert_allclose(sol_test[isol], big_sol_with_parameters(x_test, [1, 1])[0],
+                            rtol=1e-4, atol=1e-4)
```
Comment (nmayorov) [inline scipy/integrate/tests/test_bvp.py:626]: Align to the parenthesis.

**Review by nmayorov — COMMENTED** [review]
Thanks for this fix.

I've left some minor comments.

File: scipy/integrate/_bvp.py [hunk 3]
```diff
@@ -143,12 +143,15 @@ def compute_jac_indices(n, m, k):
 def stacked_matmul(a, b):
     """Stacked matrix multiply: out[i,:,:] = np.dot(a[i,:,:], b[i,:,:]).
 
-    In our case, a[i, :, :] and b[i, :, :] are always square.
+    In our case, a[i, :, :] and b[i, :, :] are square unless parameters are included in the dependent variables,
```
[the remark this answers is not in the document]
Comment (fetorres) [reply scipy/integrate/_bvp.py]: Thank you for the review. This comment is now deleted with the latest commit

File: scipy/integrate/tests/test_bvp.py [hunk 4]
```diff
@@ -152,6 +152,77 @@ def big_sol(x, n):
     return x
 
 
+def big_fun_with_parameters(x, y, p):
+    """
+    big version of sl_fun, with two parameters.
```
[the remark this answers is not in the document]
Comment (fetorres) [reply scipy/integrate/tests/test_bvp.py]: Done, thank you :)

File: scipy/integrate/tests/test_bvp.py [hunk 5]
```diff
@@ -152,6 +152,77 @@ def big_sol(x, n):
...
+
+
+def big_bc_with_parameters_jac(ya, yb, p):
+    # big version of sl_bc_jac, with two parameters
+    n = ya.shape[0]
+    dbc_dya = np.zeros((n + 2, n))
+    dbc_dyb = np.zeros((n + 2, n))
+
+    n_by_2 = int(n / 2)
+    dbc_dya[range(n_by_2), range(0, n, 2)] = 1
```
[the remark this answers is not in the document]
Comment (fetorres) [reply scipy/integrate/tests/test_bvp.py]: Done, thank you.

File: scipy/integrate/tests/test_bvp.py [hunk 6]
```diff
@@ -152,6 +152,77 @@ def big_sol(x, n):
...
+    return np.hstack((ya[::2], yb[::2], ya[1] - p[0], ya[3] - p[1]))
+
+
+def big_bc_with_parameters_jac(ya, yb, p):
+    # big version of sl_bc_jac, with two parameters
+    n = ya.shape[0]
+    dbc_dya = np.zeros((n + 2, n))
+    dbc_dyb = np.zeros((n + 2, n))
+
+    n_by_2 = int(n / 2)
```
[the remark this answers is not in the document]
Comment (fetorres) [reply scipy/integrate/tests/test_bvp.py]: OK. I made the change as suggested in your next comment :)

File: scipy/integrate/tests/test_bvp.py (hunk 1 above)
Comment (fetorres) [reply scipy/integrate/tests/test_bvp.py:623]: I see your point. What I did was mirror the tests in the test_with_params() method. I figured there might have been some subtle reason these tests were done that way for small problems :)

File: scipy/integrate/tests/test_bvp.py (hunk 2 above)
Comment (fetorres) [reply scipy/integrate/tests/test_bvp.py]: done

File: scipy/integrate/tests/test_bvp.py [hunk 7]
```diff
@@ -152,6 +152,76 @@ def big_sol(x, n):
     return x
 
 
+def big_fun_with_parameters(x, y, p):
+    """ Big version of sl_fun, with two parameters.
+
+    The two differential equations represented by sl_fun are broadcast to the number of rows of y, rotating between the
```
Comment (nmayorov) [inline scipy/integrate/tests/test_bvp.py:158]: Sorry to mention that, but it clearly violates 80 lines margin. We still adhere to it, especially when it doesn't hurt readability too much.

File: scipy/integrate/tests/test_bvp.py [hunk 8]
```diff
@@ -532,6 +602,41 @@ def test_big_problem():
...
+
+            assert_equal(sol.status, 0)
+            assert_(sol.success)
+
+            assert_allclose(sol.p, [1, 1], rtol=1e-4)
+
+            sol_test = sol.sol(x_test)
+
+            for isol in range(0, n, 4):
+                assert_allclose(sol_test[isol], big_sol_with_parameters(x_test, [1, 1])[0],
```
Comment (nmayorov) [inline scipy/integrate/tests/test_bvp.py:624]: Also seem to violate 80 lines. Consider splitting into 3 lines.

(Check other places ideally).

**Review by nmayorov — APPROVED** [review]
Consider fixing 80 lines margin violations.

File: scipy/integrate/tests/test_bvp.py (hunk 7 above)
Comment (fetorres) [reply scipy/integrate/tests/test_bvp.py]: Not a problem.

File: scipy/integrate/tests/test_bvp.py (hunk 8 above)
Comment (fetorres) [reply scipy/integrate/tests/test_bvp.py]: Done. I located all lines >80 and changed them, there were 4 lines in all. (There were 2 other lines >80 that pre-existed, i.e. were not my code. I didn't change those because I felt that would be scope creep of this PR. If you thought it worthwhile, I could do a separate PR to bring all lines in scipy.integrate within 80 characters.)

**Comment by nmayorov** [comment]
All looks correct now, merging. Thank you for the contribution!
